### ISSUE TITLE
Eliminate API handler code duplication with shared utilities

### ISSUE THREAD

**Issue by rjwalters** [rulehunt/rulehunt#101]
## Summary
Extract duplicated code from Cloudflare Pages API handlers into shared utility functions to reduce maintenance burden and improve consistency.

### Code Duplication Found

Across 4 API endpoint files (`functions/api/*.ts`), the following patterns are duplicated **identically** in each file:

#### 1. JSON Response Helper (4× duplicated)
```typescript
// Duplicated in: statistics.ts, leaderboard.ts, save.ts, starred.ts
const json = (data: unknown, status = 200) =>
  new Response(JSON.stringify(data, null, 2), {
    status,
    headers: { 'Content-Type': 'application/json' },
  })
```
**Lines of duplication**: 5 lines × 4 files = **20 lines**

#### 2. Error Handling Pattern (3× duplicated)
```typescript
// Duplicated in: statistics.ts, leaderboard.ts, starred.ts
catch (error) {
  if (error instanceof z.ZodError) {
    console.error('<Context> validation error:', error.issues)
    return json(
      { ok: false, error: 'Invalid data format', details: error.issues },
      500,
    )
  }

  if (error instanceof Error && /D1|SQL|prepare|bind/i.test(error.message)) {
    console.error('Database error <context>:', error)
    return json({ ok: false, error: 'Database query failed' }, 500)
  }

  console.error('Unexpected error in <context>:', error)
  return json({ ok: false, error: 'Internal server error' }, 500)
}
```
**Lines of duplication**: ~15 lines × 3 files = **45 lines**

### Current State
- **Total Files**: 4 API endpoints (611 total lines)
- **Duplicated Code**: ~65 lines of identical code
- **Maintenance Risk**: Changes to error handling or response format must be applied 4× independently
- **Inconsistency Risk**: `save.ts` already diverges (uses `400` status for validation errors vs `500` in others)

#### Problems with Current Approach
❌ **DRY Violation**: Same code repeated 4 times  
❌ **Maintenance Burden**: Updates require editing 4 files  
❌ **Inconsistency**: Already diverging (status codes differ between endpoints)  
❌ **Testing Gap**: Each duplication must be tested independently  
❌ **Cognitive Load**: Developers must remember to update all copies  

#### Benefits of Refactoring
✅ **Single Source of Truth**: Update once, apply everywhere  
✅ **Consistency**: Guaranteed identical behavior across endpoints  
✅ **Testability**: Test utilities once with comprehensive coverage  
✅ **Reduced LOC**: Remove ~65 lines of duplication  
✅ **Maintainability**: Future endpoint creation is faster and more reliable  

### Create Shared Utility Module

**File**: `functions/utils/api-helpers.ts`

```typescript
/// <reference types="@cloudflare/workers-types" />
import { z } from 'zod'

/**
 * Create JSON response with proper headers
 */
export function jsonResponse(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data, null, 2), {
    status,
    headers: { 'Content-Type': 'application/json' },
  })
}

/**
 * Standard error handler for API endpoints
 * Handles Zod validation errors, D1 database errors, and unexpected errors
 */
export function handleApiError(error: unknown, context: string): Response {
  // Zod validation errors
  if (error instanceof z.ZodError) {
    console.error(`${context} validation error:`, error.issues)
    return jsonResponse(
      { ok: false, error: 'Invalid data format', details: error.issues },
      400, // Use 400 for validation errors (client error)
    )
  }

  // D1 database errors
  if (error instanceof Error && /D1|SQL|prepare|bind/i.test(error.message)) {
    console.error(`Database error in ${context}:`, error)
    return jsonResponse({ ok: false, error: 'Database query failed' }, 500)
  }

  // Unexpected errors
  console.error(`Unexpected error in ${context}:`, error)
  return jsonResponse({ ok: false, error: 'Internal server error' }, 500)
}
```

### Refactor API Endpoints

**Before** (statistics.ts - 292 lines):
```typescript
export const onRequestGet = async (ctx: EventContext<...>): Promise<Response> => {
  const json = (data: unknown, status = 200) =>
    new Response(JSON.stringify(data, null, 2), { ... })

  try {
    // ... endpoint logic ...
  } catch (error) {
    if (error instanceof z.ZodError) { ... }
    if (error instanceof Error && /D1|SQL/ ...) { ... }
    console.error('Unexpected error...', error)
    return json({ ok: false, error: 'Internal server error' }, 500)
  }
}
```

**After** (statistics.ts - ~272 lines):
```typescript
import { jsonResponse, handleApiError } from '../utils/api-helpers'

export const onRequestGet = async (ctx: EventContext<...>): Promise<Response> => {
  try {
    // ... endpoint logic ...
    return jsonResponse(response)
  } catch (error) {
    return handleApiError(error, 'statistics')
  }
}
```

### Phase 1: Create Shared Utilities
1. Create `functions/utils/api-helpers.ts`
2. Implement `jsonResponse()` helper
3. Implement `handleApiError()` helper
4. Add unit tests for error handling logic

### Phase 2: Refactor Existing Endpoints (One PR Each)
1. Refactor `save.ts` (fix status code inconsistency)
2. Refactor `starred.ts`
3. Refactor `leaderboard.ts`
4. Refactor `statistics.ts`

### Phase 3: Documentation
1. Document utility usage in CLAUDE.md
2. Add JSDoc examples to utilities
3. Create migration guide for future endpoints

## Metrics

**Before**:
- Total API lines: 611
- Duplicated lines: ~65
- Files to update for error handling change: 4

**After**:
- Total API lines: ~546 (-65, -10.6%)
- Duplicated lines: 0
- Files to update for error handling change: 1

### Unit Tests for Utilities
```typescript
// functions/utils/api-helpers.test.ts
describe('jsonResponse', () => {
  it('should create JSON response with 200 status by default')
  it('should create JSON response with custom status')
  it('should set correct Content-Type header')
})

describe('handleApiError', () => {
  it('should handle Zod validation errors with 400 status')
  it('should handle D1 database errors with 500 status')
  it('should handle unexpected errors with 500 status')
  it('should log errors with context')
})
```

### Integration Testing
- Verify each refactored endpoint maintains identical behavior
- Test error scenarios (validation, database, unexpected)
- Confirm response formats unchanged

## Migration Example

**File**: `functions/api/statistics.ts`

```diff
+ import { jsonResponse, handleApiError } from '../utils/api-helpers'

  export const onRequestGet = async (ctx: EventContext<...>): Promise<Response> => {
-   const json = (data: unknown, status = 200) =>
-     new Response(JSON.stringify(data, null, 2), {
-       status,
-       headers: { 'Content-Type': 'application/json' },
-     })
-
    try {
      // ... existing logic ...
-     return json(validated)
+     return jsonResponse(validated)
    } catch (error) {
-     if (error instanceof z.ZodError) {
-       console.error('Statistics validation error:', error.issues)
-       return json(
-         { ok: false, error: 'Invalid data format', details: error.issues },
-         500,
-       )
-     }
-     if (error instanceof Error && /D1|SQL|prepare|bind/i.test(error.message)) {
-       console.error('Database error fetching statistics:', error)
-       return json({ ok: false, error: 'Database query failed' }, 500)
-     }
-     console.error('Unexpected error in statistics:', error)
-     return json({ ok: false, error: 'Internal server error' }, 500)
+     return handleApiError(error, 'statistics')
    }
  }
```

## Risk Assessment

**Low Risk** - This is a pure refactoring:
- ✅ No behavior changes (except fixing status code inconsistency in `save.ts`)
- ✅ Easy to verify with integration tests
- ✅ Can be done incrementally (one endpoint at a time)
- ✅ Rollback is trivial (git revert)

## Future Benefits

Once utilities are in place:
1. **New Endpoints**: Faster to create, guaranteed consistency
2. **Error Handling Evolution**: Can enhance centrally (add error IDs, structured logging, etc.)
3. **Testing**: Comprehensive error handling coverage in one place
4. **Monitoring**: Easy to add metrics/tracing to all endpoints

## Related Files
- `functions/api/save.ts` - 122 lines
- `functions/api/starred.ts` - 85 lines
- `functions/api/leaderboard.ts` - 112 lines
- `functions/api/statistics.ts` - 292 lines
- `functions/utils/` (to be created)

## Category
**Code Simplification** - DRY violation remediation identified by Hermit role

---

**Note**: This refactoring will reduce codebase size by ~10% in API layer while improving maintainability and consistency. The identical duplication across files indicates this was unintentional copy-paste programming that should be consolidated.

**Comment by rjwalters** [comment]
## Curator Enhancement

This issue is already **excellently specified** with comprehensive analysis and implementation plan. Adding additional technical context to make it fully ready for implementation:

---

### ✅ Acceptance Criteria

**Phase 1: Shared Utilities**
- [ ] `functions/utils/api-helpers.ts` created with TypeScript strict mode compliance
- [ ] `jsonResponse()` helper matches existing behavior exactly
- [ ] `handleApiError()` properly categorizes all error types
- [ ] Unit tests achieve 100% coverage of error handling paths
- [ ] JSDoc documentation includes usage examples

**Phase 2: Endpoint Refactoring**
For each endpoint (`save.ts`, `starred.ts`, `leaderboard.ts`, `statistics.ts`):
- [ ] Import shared utilities
- [ ] Replace inline `json()` with `jsonResponse()`
- [ ] Replace try/catch with `handleApiError()`
- [ ] Fix status code inconsistency (`save.ts` uses 400 for validation, others use 500)
- [ ] Integration tests confirm identical response format
- [ ] No regressions in error handling behavior

**Phase 3: Documentation**
- [ ] CLAUDE.md updated with utility usage guidelines
- [ ] Migration guide for future endpoints documented
- [ ] Code comments explain error categorization logic

---

### 🔍 Edge Cases & Considerations

**1. Error Type Detection**
- **Edge case**: What if an error is both a ZodError AND matches the D1 regex?
  - **Resolution**: Order matters - check ZodError first (more specific)
  
**2. Context Parameter**
- **Edge case**: Empty or undefined context string
  - **Resolution**: Use default context or validate input

**3. Error Logging**
- **Consideration**: Should we log stack traces for unexpected errors?
  - **Recommendation**: Yes, include `error.stack` in console.error for debugging

**4. Response Format Consistency**
- **Critical**: Ensure `details` field is only present for Zod errors
  - **Test**: Verify D1 errors don't leak internal details to client

**5. Status Code Standardization**
- **Decision needed**: Should validation errors be 400 or 500?
  - **Recommendation**: Use 400 (client error) per HTTP spec
  - **Impact**: This fixes existing inconsistency in `save.ts`

---

### 🧪 Additional Test Cases

```typescript
// functions/utils/api-helpers.test.ts

describe('jsonResponse', () => {
  it('should handle null data')
  it('should handle undefined data')
  it('should handle large objects without truncation')
  it('should format JSON with 2-space indentation')
})

describe('handleApiError', () => {
  // Zod errors
  it('should extract all Zod validation issues')
  it('should handle nested Zod schema errors')
  
  // D1 errors
  it('should detect D1_ERROR in message')
  it('should detect SQL syntax errors')
  it('should detect prepare/bind failures')
  
  // Unknown errors
  it('should handle string errors (thrown as strings)')
  it('should handle null/undefined errors')
  it('should handle errors without message property')
  
  // Context
  it('should include context in all log messages')
  it('should sanitize context to prevent log injection')
})
```

---

### 📊 Verification Steps

After each endpoint refactoring:

1. **Response Format Test**:
   ```bash
   # Before and after should be identical
   curl https://rulehunt.org/api/statistics | jq '.' > before.json
   # (deploy changes)
   curl https://rulehunt.org/api/statistics | jq '.' > after.json
   diff before.json after.json  # Should be empty
   ```

2. **Error Handling Test**:
   ```bash
   # Test validation error (should return 400 with details)
   curl -X POST https://rulehunt.org/api/save -d '{invalid}'
   
   # Test database error (should return 500 without internal details)
   # (simulate by breaking D1 binding)
   ```

3. **Performance Test**:
   - Verify response time unchanged (utilities should add negligible overhead)
   - No memory leaks from error handler closures

---

### 🛡️ Safety Considerations

**1. Incremental Migration**
- ✅ Refactor one endpoint per PR
- ✅ Run full integration tests after each
- ✅ Monitor production logs for unexpected errors

**2. Rollback Plan**
- Each PR is independently revertible
- Utility module can coexist with old code during migration
- No database schema changes required

**3. Error Information Leakage**
- ❌ Never expose D1 internal errors to clients
- ❌ Never include stack traces in API responses
- ✅ Log full details server-side for debugging
- ✅ Return generic messages to clients

---

### 🎯 Success Metrics

**Code Quality**:
- Lines of duplicated code: 65 → 0
- Code coverage in API layer: +15%
- Linting errors: 0

**Maintainability**:
- Time to add new endpoint: -30% (estimated)
- Files to update for error format change: 4 → 1

**Consistency**:
- Status code inconsistencies: 1 → 0
- Error response format variations: Multiple → Single

---

### 🚀 Ready for Implementation

This issue is now **fully specified** and ready for the Builder role to implement. The existing analysis is excellent, and these additions provide the final details needed for confident implementation.

**Recommended Approach**: Start with Phase 1 (utilities + tests), then refactor endpoints one at a time with thorough testing between each.

**Estimated Effort**: 
- Phase 1: 2-3 hours (utilities + comprehensive tests)
- Phase 2: 1 hour per endpoint × 4 = 4 hours
- Phase 3: 1 hour (documentation)
- **Total**: ~7-8 hours

🤖 Curated with [Claude Code](https://claude.com/claude-code)
